test/integration/util/shell.go

```diff
@@ -0,0 +1,278 @@
+// +build integration
+
+/*
+Copyright (C) 2017 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"github.com/DATA-DOG/godog/gherkin"
+	"github.com/minishift/minishift/pkg/util/os"
+)
+
+const (
+	exitCodeIdentifier = "exitCodeOfLastCommandInShell="
+
+	bashExitCodeCheck       = "echo %v$?"
+	fishExitCodeCheck       = "echo %v$status"
+	tcshExitCodeCheck       = "echo %v$?"
+	zshExitCodeCheck        = "echo %v$?"
+	cmdExitCodeCheck        = "echo %v%%errorlevel%%"
+	powershellExitCodeCheck = "echo %v$lastexitcode"
+)
+
+var (
+	shellStartArgument    []string
+	shellName             string
+	shellCheckExitCodeCmd string
+
+	minishiftPath string
+
+	shell  *exec.Cmd
+	outbuf bytes.Buffer
+	errbuf bytes.Buffer
+	excbuf bytes.Buffer
+
+	outPipe io.ReadCloser
+	errPipe io.ReadCloser
+	inPipe  io.WriteCloser
+
+	outScanner *bufio.Scanner
+	errScanner *bufio.Scanner
+
+	stdoutChannel   chan string
+	stderrChannel   chan string
+	exitCodeChannel chan string
+)
+
+func SetShell(name string, mpath string) {
+	shellName = name
+	minishiftPath = mpath
+}
+
+func configureShell() {
+	switch shellName {
+	case "bash":
+		shellCheckExitCodeCmd = fmt.Sprintf(bashExitCodeCheck, exitCodeIdentifier)
+	case "tcsh":
+		shellCheckExitCodeCmd = fmt.Sprintf(tcshExitCodeCheck, exitCodeIdentifier)
+	case "zsh":
+		shellCheckExitCodeCmd = fmt.Sprintf(zshExitCodeCheck, exitCodeIdentifier)
+	case "cmd":
+		shellCheckExitCodeCmd = fmt.Sprintf(cmdExitCodeCheck, exitCodeIdentifier)
+	case "powershell":
+		shellStartArgument = []string{"-Command", "-"}
+		shellCheckExitCodeCmd = fmt.Sprintf(powershellExitCodeCheck, exitCodeIdentifier)
+	case "fish":
+		fmt.Println("Fish shell is currently not supported by integration tests. Default shell for the OS will be used.")
+		shellName = ""
+		fallthrough
+	default:
+		if shellName != "" {
+			fmt.Printf("Shell %v is not supported, will set the default shell for the OS to be used.\n", shellName)
+		}
+		switch os.CurrentOS() {
+		case "darwin", "linux":
+			shellName = "bash"
+			shellCheckExitCodeCmd = fmt.Sprintf(bashExitCodeCheck, exitCodeIdentifier)
+		case "windows":
+			shellName = "powershell"
+			shellStartArgument = []string{"-Command", "-"}
+			shellCheckExitCodeCmd = fmt.Sprintf(powershellExitCodeCheck, exitCodeIdentifier)
+		}
+	}
+
+	return
+}
+
+func StartShellInstance() error {
+	var err error
```

Comment:
it would be nicer to encapsulate the shell functionality into to struct

Reply:
@hferentschik Will take a look on that and implement it. Thanks.